lib/SILOptimizer/Transforms/SpecializeOpaqueArchetypes.cpp

```diff
@@ -0,0 +1,155 @@
+  }
+
+  void visitTerminator(SILBasicBlock *BB) { visit(BB->getTerminator()); }
+
+  SILType remapType(SILType Ty) {
+    SILType &Sty = TypeCache[Ty];
+    if (!Sty) {
+      // Apply the opaque types substitution.
+      ReplaceOpaqueTypesWithUnderlyingTypes replacer;
+      Sty = Ty.subst(Original.getModule(), SubsMap)
+                .subst(Original.getModule(), replacer, replacer,
+                       CanGenericSignature(), true);
+    }
+    return Sty;
+  }
+
+  CanType remapASTType(CanType ty) {
+    // Apply the opaque types substitution.
+    ReplaceOpaqueTypesWithUnderlyingTypes replacer;
+    return SuperTy::remapASTType(ty)
+        .subst(replacer, replacer,
+               SubstFlags::SubstituteOpaqueArchetypes |
+                   SubstFlags::AllowLoweredTypes)
+        ->getCanonicalType();
+  }
+
+  ProtocolConformanceRef remapConformance(Type type,
+                                          ProtocolConformanceRef conf) {
+    // Apply the opaque types substitution.
+    ReplaceOpaqueTypesWithUnderlyingTypes replacer;
+    return SuperTy::remapConformance(type, conf)
+        .subst(type, replacer, replacer,
+               SubstFlags::SubstituteOpaqueArchetypes |
+                   SubstFlags::AllowLoweredTypes);
+  }
+
+  SubstitutionMap remapSubstitutionMap(SubstitutionMap Subs) {
+    // Apply the opaque types substitution.
+    ReplaceOpaqueTypesWithUnderlyingTypes replacer;
+    return SuperTy::remapSubstitutionMap(Subs).subst(
+        replacer, replacer,
+        SubstFlags::SubstituteOpaqueArchetypes | SubstFlags::AllowLoweredTypes);
+  }
+};
+} // namespace
+
+void OpaqueSpecializerCloner::clone() {
+  for (auto arg: entryBlock->getArguments())
+    recordFoldedValue(arg, arg);
+  cloneReachableBlocks(cloneFromBlock, {}, entryBlock,
+                       true /*havePrepopulatedFunctionArgs*/);
+  getBuilder().setInsertionPoint(entryBlock);
+  getBuilder().createBranch(RegularLocation::getAutoGeneratedLocation(),
+                            getOpBasicBlock(cloneFromBlock));
+}
+
+/// Update address uses of the opaque type archetype with the concrete type.
+/// This is neccessary for apply instructions.
+void OpaqueSpecializerCloner::insertOpaqueToConcreteAddressCasts(
+    SILInstruction *orig, SILInstruction *cloned) {
+
+  // Replace apply operands.
+  auto apply = ApplySite::isa(cloned);
+  if (!apply)
+    return;
+
+  SavedInsertionPointRAII restore(getBuilder());
+  getBuilder().setInsertionPoint(apply.getInstruction());
+  auto substConv = apply.getSubstCalleeConv();
+  unsigned idx = 0;
+  for (auto &opd : apply.getArgumentOperands()) {
+    auto argConv = apply.getArgumentConvention(opd);
+    auto argIdx = apply.getCalleeArgIndex(opd);
+    auto argType = substConv.getSILArgumentType(argIdx);
+    if (argConv.isIndirectConvention() &&
+        argType.getASTType()->hasOpaqueArchetype() &&
+        !opd.get()->getType().getASTType()->hasOpaqueArchetype()) {
```

Comment:
Maybe just check that the types differ? What if both have opaque types but they're different, because of various inlining and specialization passes?